src/components/Figure/Image.js

```diff
@@ -1,20 +1,91 @@
-import React from 'react'
+import React, { Component } from 'react'
 import PropTypes from 'prop-types'
-import { css } from 'glamor'
+import { css, merge } from 'glamor'
+import { parse, format } from 'url'
 
 const styles = {
   image: css({
     width: '100%'
+  }),
+  imageLoading: css({
+    backgroundColor: '#eee',
+    height: 0
   })
 }
 
-const Image = ({ src, alt, attributes = {} }) => {
-  return <img {...attributes} {...styles.image} src={src} alt={alt} />
+const imageSizeInfo = url => {
+  const urlObject = parse(url, true)
+  const { size } = urlObject.query
+  if (!size) {
+    return null
+  }
+  const [width, height] = size.split('x')
+  return {
+    width,
+    height
+  }
+}
+
+const imageResizeUrl = (url, size) => {
+  if (!url) {
+    return url
+  }
+
+  const urlObject = parse(url, true)
+  if (urlObject.protocol === 'data:') {
+    return url
+  }
+
+  urlObject.query.resize = size
+  // ensure format calculates from query object
+  urlObject.search = undefined
+
+  return format(urlObject)
+}
+
+class Image extends Component {
+  constructor(props) {
+    super(props)
+    this.state = { loadStatus: 'loading' }
+
+    this.handleLoaded = this.handleLoaded.bind(this)
+  }
+
+  handleLoaded() {
+    this.setState({ loadStatus: 'loaded' })
+  }
+
+  render() {
+    const { src, alt, attributes = {}, resize } = this.props
+    const { loadStatus } = this.state
+    const resizedSrc = resize ? imageResizeUrl(src, resize) : src
+    const sizeInfo = imageSizeInfo(src)
+    const aspectRatio =
+      sizeInfo && sizeInfo.width ? sizeInfo.width / sizeInfo.height : undefined
+
+    return (
+      <img
+        {...attributes}
+        {...css(
+          styles.image,
+          aspectRatio && loadStatus === 'loading'
+            ? merge(styles.imageLoading, {
+                paddingBottom: `${100 / aspectRatio}%`
+              })
+            : {}
+        )}
+        onLoad={this.handleLoaded}
```

Comment:
`onLoad` is not supper reliable and behaves differently depending in the browser engine, e.g. only the first time a image loads:
https://bugs.chromium.org/p/chromium/issues/detail?id=7731#c12

I'd prefer to go with a stateless solution:
```
<span {...styles.aspectRadio} style={{paddingBottom: `...`}}>
  <img />
</span>
```

A `span` with `display:block;` images are inline blocks and can be used where `div` aren't strictly allowed.

Alternative would be to handle changing `src` props, add `onError` handle and put a `key={src}` on the image to tell react to not re-use the dom node when src changes.

Reply:
I've been thinking about this some more: our images will become stateful for other reasons—proper lazy loading on scroll, srcset fallback for old browser or a polyfill implementation—don't spend any time making it stateless just for the sack of it. The `onLoad` concern are still there though.